### ISSUE TITLE
[REQ] Add Description to generators

### ISSUE THREAD

**Issue by DanielHabenicht** [OpenAPITools/openapi-generator#3219]
### Is your feature request related to a problem? Please describe.

When browsing the [available generators](https://openapi-generator.tech/docs/generators/ada) and navigating to the detail page no description of the generator is given (or information on how to use it in your application).

## Describe the solution you'd like

Add some information to these pages, especially: 
 - some basic information on what is generated
 - Requirements that should be fulfilled to use the code generated
 - how it should be used in your application
 - how it can not be used and possible workarounds
 - (Maybe an example command to generate it, although that's covered by the general guide, but it makes life easier to just copy a command)

I am highly motivated to add this information to the generators listed below but would welcome a discussion about the structure or any other information missing. 

## Additional context

While looking for a generator for my new node-js backend I tried several generators, namely 
 - `typescript-node`, which I could not get to work, because I was to lazy to find out which other packages I would need to install (this isn't mentioned anywhere)
 - `typescript-rxjs`, which does not work out-of-the-box with node-js
 - `typescript-axios`, which finally worked because the name already says what to install :)

**Comment by macjohnny** [comment]
Sounds good. Would it be enough to add a link to a generated Readme of the corresponding generator to the docs page?

**Comment by DanielHabenicht** [comment]
I like the idea, but unfortunately, not all generators do have a `readme.md`. 
Is there a repo with the current output of the generators to link to?

**Comment by macjohnny** [comment]
e.g. https://github.com/OpenAPITools/openapi-generator/blob/master/samples/client/petstore/typescript-angular-v7-provided-in-root/builds/with-npm/README.md

It might be worth adding Readme files to the generators

**Comment by macjohnny** [comment]
@jimschubert what do you think? looking for the generated Readme to display it in https://openapi-generator.tech/docs/generators/ada?

**Comment by DanielHabenicht** [comment]
> It might be worth adding Readme files to the generators

It sure is. I am looking for a way to make it as easy as possible to add content to it, but: 
 - not every generator has an example project like the one mentioned (this is a requirement for your solution)
 - not everybody is familiar with the template syntax to add something to the readme. 
 - furthermore there is no obvious link from the [generator-detail-site](https://openapi-generator.tech/docs/generators/ada) to the readme.md that has to be edited. 

I think there are two option: 
1. add it to the readme of the generator that does already exist (but maybe creating duplicate data)
2. generate an example project for every project and link the readme to generator page. 

I really like the second idea, but maybe that's to complex. Although it has some great side effects like being able to see what the generated source code will look like without having to write one command.

**Comment by macjohnny** [comment]
there should be an example project for every generator, however, not every generator has a readme.

the template syntax does not have to be used, the `Readme.mustache` can contain markdown which will be directly output to ``Readme.md`, so no problem either.

currently, there is no link from the generator detail site, that's true, but I think this could be added automatically

**Comment by DanielHabenicht** [comment]
> there should be an example project for every generator, however, not every generator has a readme.

Ahh, didn't saw them on the first look. Are they kind of unorganized, because it looks like there are duplicates: e.g. [server/openapi3](https://github.com/OpenAPITools/openapi-generator/tree/master/samples/server/openapi3/petstore) and [openapi3/server](https://github.com/OpenAPITools/openapi-generator/tree/master/samples/openapi3/server/petstore))?

> the template syntax does not have to be used, the `Readme.mustache` can contain markdown which will be directly output to ``Readme.md`, so no problem either.
> 
> currently, there is no link from the generator detail site, that's true, but I think this could be added automatically

Looks like this makes it even easier. I will go ahead and create a Readme in the repo where it was missing. 

Also is there a guide to write new generators? (Everything I found was the [CONTRIBUTING.md](https://github.com/OpenAPITools/openapi-generator/blob/master/CONTRIBUTING.md)) It would be nice to have a checklist where `Write a Readme.md` could be added.

**Comment by jimschubert** [comment]
@DanielHabenicht See https://openapi-generator.tech/docs/new-generator

If the steps here are followed, there will be a generated README, we however don't have rules about what goes into the README.

I don't think it would be too helpful to add the README to the doc site, which is intended to document openapi-generator and not it's outputs. I do think we should have a quality gate around terms or headers required for built-in generators.

**Comment by DanielHabenicht** [comment]
> @DanielHabenicht See https://openapi-generator.tech/docs/new-generator

Neat, I might add a link to the CONTRIBUTING.md

> I don't think it would be too helpful to add the README to the doc site, which is intended to document openapi-generator and not it's outputs. 

In my mind, there should be some information about what is generated, how it can be used, and what requirements it has before generating it. Otherwise, a user would have to generate each generator before choosing one, in order to read the readme.

**Comment by macjohnny** [comment]
@jimschubert 
> I don't think it would be too helpful to add the README to the doc site

I disagree, since the readme contains information about how to use the generated code, this will help to understand whether the code generator is useful for ones needs and what to expect to be generated. Moreover, a link to the generated samples might be useful as well.

If we had a link to the Readme, it might animate contributors to add more information to the readme to better understand the concept/intended usage.

**Comment by jimschubert** [comment]
@macjohnny the usage of the generated code will change depending on combinations of input parameters, though. It's apparently different across versions, and the docsite is currently only the latest released version.

I think there's value in having a feature matrix in the doc site (there's an issue to propose that work), but I think it would be overwhelming to have multiple README versions per generator and perhaps more confusing than not having them.

We do mention the `samples` directory, where a subset of options and generated output have READMEs viewable in Markdown. Maybe an option would be to link off to the samples rather than duplicating the READMEs on the doc site?

**Comment by DanielHabenicht** [comment]
> @macjohnny the usage of the generated code will change depending on combinations of input parameters, though. It's potentially different across versions, and the docsite is currently only the latest released version.
> I think there's value in having a feature matrix in the doc site (there's an issue to propose that work), but I think it would be overwhelming to have multiple README versions per generator and perhaps more confusing than not having them.

I agree, but maybe we are talking past another. 
1. Every generator should have `Readme.md`, that's not the case for every generator but something that should be addressed in a different issue. (It solves a part of my problem and I updated my initial issue comment to reflect this)
2. General Usage of the generator is not described (and that's what I am missing in the detail page), to clarify what I mean, e.g. for the [typescript-rxjs](https://openapi-generator.tech/docs/generators/typescript-rxjs) generator page:

> This is the typescript-rxjs generator. It is of great use for browser based application, e.g. Angular, React, Vue. 
> It does support caching out of the box and feature 1 aswell as feature 2.  
>
>Generate it with: `npx openapi-generator generate -i /path/to/your/openapi.yaml -g typescript-rxjs -o /output/path/`
>
>It is currently not supporting Applications using Node. If you want to use it with node see this workaround: #2731
> Have a look at the `latest` generated [Readme.md](link) to learn more. 

This does save developers time by not having to search for the Readme but also getting a general overview of the generator without generating it. 

After explaining the impact of the options and different version on the Readme I agree that including it might confuse. I support to link to latest readme.
